odoo/models.py

```diff
@@ -7468,83 +7556,189 @@ def itemgetter_tuple(items):
-        return {'message': env._("Missing required value for the field '%(name)s' on a linked model [%(linked_model)s]", name=e.diag.column_name, linked_model=e.diag.table_name)}
-
-    field_name = e.diag.column_name
-    field = fields[field_name]
-    message = env._("Missing required value for the field '%(name)s' (%(technical_name)s)", name=field['string'], technical_name=field_name)
-    return {
-        'message': message,
-        'field': field_name,
-    }
-
-
-def convert_pgerror_unique(model, fields, info, e):
-    # new cursor since we're probably in an error handler in a blown
-    # transaction which may not have been rollbacked/cleaned yet
-    with closing(model.env.registry.cursor()) as cr_tmp:
-        cr_tmp.execute(SQL("""
-            SELECT
-                conname AS "constraint name",
-                t.relname AS "table name",
-                ARRAY(
-                    SELECT attname FROM pg_attribute
-                    WHERE attrelid = conrelid
-                      AND attnum = ANY(conkey)
-                ) as "columns"
-            FROM pg_constraint
-            JOIN pg_class t ON t.oid = conrelid
-            WHERE conname = %s
-        """, e.diag.constraint_name))
-        constraint, table, ufields = cr_tmp.fetchone() or (None, None, None)
-    # if the unique constraint is on an expression or on an other table
-    if not ufields or model._table != table:
-        return {'message': tools.exception_to_unicode(e)}
-
-    # TODO: add stuff from e.diag.message_hint? provides details about the constraint & duplication values but may be localized...
-    if len(ufields) == 1:
-        field_name = ufields[0]
-        field = fields[field_name]
-        message = model.env._(
-            "The value for the field '%(field)s' already exists (this is probably '%(other_field)s' in the current model).",
-            field=field_name,
-            other_field=field['string'],
+class SQLObject:
+    """ Declares a SQL object related to the model.
+
+    The name of the SQL object will be "{model._table}_{key}".
+    """
+    key: str
+    message: ConstraintMessageType = ''
+
+    def __init__(self):
+        """Abstract SQL object"""
+        self.key = ''
+
+    def __set_name__(self, owner, name):
+        assert isinstance(owner, MetaModel)
+        assert name.startswith('_'), "Names of SQL objects in a model must start with '_'"
+        self.key = name[1:]
+        owner._local_sql_objects[self.key] = self
+
+    @property
+    def definition(self) -> str:
+        raise NotImplementedError
+
+    def full_name(self, model: BaseModel) -> str:
+        assert self.key, f"The SQL object is not named ({self.definition})"
+        name = f"{model._table}_{self.key}"
+        return sql.make_identifier(name)
+
+    def _set_message(self, message: ConstraintMessageType):
+        if isinstance(message, str):
+            _logger.debug("Use a lazy translation or a function message for %r", self)
```

Comment:
Is it an error ? Do you always want lazy translation  or never ?

Debug log is quite pointless IMO.